crypto/tbs/src/lib.rs

```diff
@@ -174,40 +175,49 @@ pub fn sign_blinded_msg(msg: BlindedMessage, sks: SecretKeyShare) -> BlindedSign
     BlindedSignatureShare(sig.to_affine())
 }
 
-/// Combines a sufficient amount of valid blinded signature shares to a blinded
-/// signature. The responsibility of verifying the supplied shares lies with the
-/// caller.
-///
-/// * `sig_shares`: an iterator yielding pairs of key indices and signature
-///   shares from said key
-/// * `threshold`: number of shares needed to combine a signature
-///
+/// Combines the exact threshold of valid blinded signature shares to a blinded
+/// signature. The responsibility of verifying the shares lies and supplying
+/// exactly the necessary threshold of shares lies with the caller.
 /// # Panics
-/// If the amount of shares supplied is less than the necessary amount
-pub fn combine_valid_shares<I>(sig_shares: I, threshold: usize) -> BlindedSignature
-where
-    I: IntoIterator<Item = (usize, BlindedSignatureShare)>,
-    I::IntoIter: Clone + ExactSizeIterator,
-{
-    let points = sig_shares
-        .into_iter()
-        .take(threshold)
-        .map(|(idx, share)| {
-            let x = Scalar::from((idx as u64) + 1);
-            let y = share.0.into();
-            (x, y)
-        })
-        .collect::<Vec<(Scalar, G1Projective)>>();
-    if points.len() < threshold {
-        panic!("Not enough signature shares");
-    }
+/// If the amount of shares supplied is less than two
+pub fn aggregate_signature_shares(
+    shares: &BTreeMap<u64, BlindedSignatureShare>,
+) -> BlindedSignature {
+    BlindedSignature(
+        lagrange_multiplier(shares.keys().cloned().map(Scalar::from).collect())
+            .into_iter()
+            .zip(shares.values())
+            .map(|(lm, y)| lm * y.0)
+            .reduce(|a, b| a + b)
+            .expect("We have at least one share")
+            .to_affine(),
+    )
+}
 
-    if points.len() == 1 {
-        return BlindedSignature(points.first().unwrap().1.to_affine());
-    }
+pub fn aggregate_public_key_shares(shares: &BTreeMap<u64, PublicKeyShare>) -> AggregatePublicKey {
+    AggregatePublicKey(
+        lagrange_multiplier(shares.keys().cloned().map(Scalar::from).collect())
+            .into_iter()
+            .zip(shares.values())
+            .map(|(lm, y)| lm * y.0)
+            .reduce(|a, b| a + b)
+            .expect("We have at least one share")
+            .to_affine(),
```

Comment:
Feels like parts of this could be refactored out and de-duplicated xD but I guess your goal was to simplify and accept redundancy.